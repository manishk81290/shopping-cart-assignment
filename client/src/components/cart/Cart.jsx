import React, { useContext } from "react";
import { Container, Row, Col, Modal, Button } from "react-bootstrap";
import CartItem from "./CartItem";

import ShoppingContext from "../../context/ShoppingContext";

import "./cart.scss";

const Cart = ({ show, handleClose }) => {
  const shoppingContext = useContext(ShoppingContext);
  const { cart, updateCart } = shoppingContext;

  const updateQuantity = (action, index) => {
    let newCart = [...cart];
    if (action === "increment") {
      newCart[index]["count"] += 1;
    } else {
      newCart[index]["count"] -= 1;
      if (newCart[index]["count"] === 0) {
        newCart.splice(index, 1);
      }
    }
    updateCart(newCart);
  };

  return (
    <Modal show={show} onHide={handleClose}>
      <Modal.Header closeButton>
        <Modal.Title>
          My Cart{" "}
          <span>
            ({cart.reduce((accumulator, item) => accumulator + item.count, 0)}{" "}
            item)
          </span>
        </Modal.Title>
      </Modal.Header>
      <Modal.Body className={!cart.length && "no-item"}>
        <Container fluid>
          {cart.length ? (
            cart.map((item, index) => {
              return (
                <CartItem
                  item={item}
                  index={index}
                  key={item.id}
                  updateQuantity={updateQuantity}
                />
              );
            })
          ) : (
            <Row md={12}>
              <Col md={12}>
                No Item in the cart
                <span>Your favourite items are just a click away</span>
              </Col>
            </Row>
          )}
        </Container>
      </Modal.Body>
      <Modal.Footer>
        <p>Promocode can be applied on payment page</p>
        <Button
          variant="primary"
          className="theme-button w-100"
          onClick={handleClose}
        >
          <span className="btn-text">Proceed to checkout</span>
          <span className="total">
            Rs.
            {cart.reduce(
              (accumulator, item) => accumulator + item.count * item.price,
              0
            )}{" "}
            &gt;
          </span>
        </Button>
      </Modal.Footer>
    </Modal>
  );
};

export default Cart;
