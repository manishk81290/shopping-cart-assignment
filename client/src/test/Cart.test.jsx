import React from "react";
import Cart from "./../components/cart/Cart";
import Adapter from "enzyme-adapter-react-16";
import { shallow, configure } from "enzyme";

configure({ adapter: new Adapter() });
let realUseContext;
let useContextMock;
const sampleContextData = {
  cart: [
    {
      name: "Capsicum - Green, 1 kg",
      imageURL: "/static/images/products/fruit-n-veg/capsicum-green.jpg",
      description:
        "Leaving a moderately pungent taste on the tongue, Green capsicums, also known as green peppers are bell shaped, medium-sized fruit pods.",
      price: 137,
      stock: 50,
      category: "5b6899953d1a866534f516e2",
      sku: "fnw-capsicum-1",
      id: "5b6c6bdc01a7c38429530886",
      count: 2,
    },
  ],
  updateCart: () => {},
};

const sampleContextEmptyData = {
  cart: [],
  updateCart: () => {},
};

const testProps = {
  handleClose: jest.fn(),
  show: true,
};

beforeEach(() => {
  realUseContext = React.useContext;
  useContextMock = React.useContext = jest.fn();
});
// Cleanup mock
afterEach(() => {
  React.useContext = realUseContext;
});

describe("Cart Component test cases", () => {
  it("Cart should match with the snapshot", () => {
    useContextMock.mockReturnValue(sampleContextData);
    const tree = shallow(<Cart {...testProps} />);
    expect(tree).toBeTruthy();
    expect(tree).toMatchSnapshot();
  });
  it("Empty cart shouldn't render any item", () => {
    useContextMock.mockReturnValue(sampleContextEmptyData);
    const tree = shallow(<Cart {...testProps} />);
    expect(tree).toBeTruthy();
  });
});
