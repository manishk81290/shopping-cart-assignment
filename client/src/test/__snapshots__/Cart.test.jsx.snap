// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart with mock useContext hook 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop={true}
  dialogAs={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    }
  }
  enforceFocus={true}
  keyboard={true}
  onHide={[MockFunction]}
  restoreFocus={true}
  show={[Function]}
>
  <ModalHeader
    closeButton={true}
    closeLabel="Close"
  >
    <ModalTitle>
      My Cart
       
      <span>
        (
        2
         
        item)
      </span>
    </ModalTitle>
  </ModalHeader>
  <ModalBody
    className={false}
  >
    <Container
      fluid={true}
    >
      <CartItem
        index={0}
        item={
          Object {
            "category": "5b6899953d1a866534f516e2",
            "count": 2,
            "description": "Leaving a moderately pungent taste on the tongue, Green capsicums, also known as green peppers are bell shaped, medium-sized fruit pods.",
            "id": "5b6c6bdc01a7c38429530886",
            "imageURL": "/static/images/products/fruit-n-veg/capsicum-green.jpg",
            "name": "Capsicum - Green, 1 kg",
            "price": 137,
            "sku": "fnw-capsicum-1",
            "stock": 50,
          }
        }
        updateQuantity={[Function]}
      />
    </Container>
  </ModalBody>
  <ModalFooter>
    <p>
      Promocode can be applied on payment page
    </p>
    <Button
      active={false}
      className="theme-button w-100"
      disabled={false}
      onClick={[MockFunction]}
      variant="primary"
    >
      <span
        className="btn-text"
      >
        Proceed to checkout
      </span>
      <span
        className="total"
      >
        Rs.
        274
         
        &gt;
      </span>
    </Button>
  </ModalFooter>
</Modal>
`;
